examples/upstream.rs

```diff
@@ -0,0 +1,356 @@
+            (*hcpd).client_connection = Some(request.connection());
+            (*hcpd).original_get_peer = (*upstream_ptr).peer.get;
+            (*hcpd).original_free_peer = (*upstream_ptr).peer.free;
+
+            (*upstream_ptr).peer.data = hcpd as *mut c_void;
+            (*upstream_ptr).peer.get = Some(ngx_http_upstream_get_custom_peer);
+            (*upstream_ptr).peer.free = Some(ngx_http_upstream_free_custom_peer);
+        }
+
+        ngx_log_debug_http!(request, "CUSTOM UPSTREAM end request peer init");
+        Status::NGX_OK
+    }
+);
+
+// ngx_http_usptream_get_custom_peer
+// For demonstration purposes, use the original get callback, but log this callback proxies through
+// to the original.
+#[no_mangle]
+unsafe extern "C" fn ngx_http_upstream_get_custom_peer(pc: *mut ngx_peer_connection_t, data: *mut c_void) -> ngx_int_t {
+    let hcpd: *mut UpstreamPeerData = unsafe { mem::transmute(data) };
+
+    ngx_log_debug_mask!(
+        NGX_LOG_DEBUG_HTTP,
+        (*pc).log,
+        "CUSTOM UPSTREAM get peer, try: {}, conn: {:p}",
+        (*pc).tries,
+        (*hcpd).client_connection.unwrap(),
+    );
+
+    let original_get_peer = (*hcpd).original_get_peer.unwrap();
+    let rc = original_get_peer(pc, (*hcpd).data);
+
+    if rc != Status::NGX_OK.into() {
+        return rc;
+    }
+
+    ngx_log_debug_mask!(NGX_LOG_DEBUG_HTTP, (*pc).log, "CUSTOM UPSTREAM end get peer");
+    Status::NGX_OK.into()
+}
+
+// ngx_http_upstream_free_custom_peer
+// For demonstration purposes, use the original free callback, but log this callback proxies
+// through to the original.
+#[no_mangle]
+unsafe extern "C" fn ngx_http_upstream_free_custom_peer(
+    pc: *mut ngx_peer_connection_t,
+    data: *mut c_void,
+    state: ngx_uint_t,
+) {
+    ngx_log_debug_mask!(NGX_LOG_DEBUG_HTTP, (*pc).log, "CUSTOM UPSTREAM free peer");
+
+    let hcpd: *mut UpstreamPeerData = unsafe { mem::transmute(data) };
+
+    let original_free_peer = (*hcpd).original_free_peer.unwrap();
+
+    original_free_peer(pc, (*hcpd).data, state);
+
+    ngx_log_debug_mask!(NGX_LOG_DEBUG_HTTP, (*pc).log, "CUSTOM UPSTREAM end free peer");
+}
+
+// ngx_http_upstream_init_custom
+// The module's custom `peer.init_upstream` callback.
+// The original callback is saved in our SrvConfig data and reset to this module's `peer.init`.
+#[no_mangle]
+unsafe extern "C" fn ngx_http_upstream_init_custom(
+    cf: *mut ngx_conf_t,
+    us: *mut ngx_http_upstream_srv_conf_t,
+) -> ngx_int_t {
+    ngx_log_debug_mask!(NGX_LOG_DEBUG_HTTP, (*cf).log, "CUSTOM UPSTREAM peer init_upstream");
+
+    let maybe_conf: Option<*mut SrvConfig> =
+        ngx_http_conf_upstream_srv_conf_mutable(us, &ngx_http_upstream_custom_module);
+    if maybe_conf.is_none() {
+        ngx_conf_log_error(
+            NGX_LOG_EMERG as usize,
+            cf,
+            0,
+            "CUSTOM UPSTREAM no upstream srv_conf".as_bytes().as_ptr() as *const i8,
+        );
```

Comment:
nit: a good candidate to be a macro for ngx crate
ngx_conf_log_error!(LEVEL, "regular text");
seems like we need to revisit logging and understand it better from a usage point of view. Maybe for a next time.

Reply:
i'll convert this to an issue.